File: Mobile Applications/android/app/src/main/java/com/example/activityplannerapp/activities/service/ActivitiesService.kt
package com.example.activityplannerapp.activities.service

import com.example.activityplannerapp.activities.domain.Activity
import com.example.activityplannerapp.activities.repo.ActivitiesRepository
import com.example.activityplannerapp.activities.repo.ActivitiesRepositoryImpl
import com.example.activityplannerapp.activities.server.ActivityAPI.isConnectionOn
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.toList

class ActivitiesService (
    private val activityRepo: ActivitiesRepository
) {
    suspend fun getActivities() = activityRepo.getActivities()

    suspend fun getActivityById(id: Int?): Activity {
        if (isConnectionOn)
            return activityRepo.getActivityById(id)
        return activityRepo.getLocalActivityById(id)
    }

//    suspend fun checkOverlap(newActivity: Activity) {
//        var activities: List<Activity> = emptyList()
//
//        getActivities().collect {
//            activities = it
//        }
//        for (activity in activities) {
//            if (newActivity.date == activity.date && newActivity.id != activity.id) {
//                // Check if the new activity's startTime and endTime overlap with existing activity
//                if (newActivity.startTime.isBefore(activity.endTime) && newActivity.endTime.isAfter(activity.startTime)) {
//                    throw IllegalArgumentException("Activity overlaps with an existing activity")
//                }
//            }
//        }
//    }

    private fun checkInvalidTimes(newActivity: Activity) {
        if (newActivity.startTime.isAfter(newActivity.endTime)) {
            throw IllegalArgumentException("Activity's start time is after end time")
        }
    }

    suspend fun addActivity(newActivity: Activity) {
        if (newActivity.title == "") {
            throw IllegalArgumentException("Insert a title for this activity")
        }
        checkInvalidTimes(newActivity)
        // checkOverlap(newActivity)
        if (isConnectionOn) {
            activityRepo.addActivity(newActivity)
        } else {
            activityRepo.addLocalActivity(newActivity)
        }
    }

    suspend fun updateActivity(activity: Activity) {
        if (activity.title == "") {
            throw IllegalArgumentException("Insert a title for this activity")
        }
        checkInvalidTimes(activity)
        // checkOverlap(activity)
        if (isConnectionOn) {
            activityRepo.updateActivity(activity)
        } else {
            activityRepo.updateLocalActivity(activity)
        }
    }

    suspend fun deleteActivity(activity: Activity) = activityRepo.deleteActivity(activity)
}

File: Mobile Applications/flutter_app/lib/Service/ActivitiesService.dart

import 'package:flutter/material.dart';
import 'package:flutter_app/Repo/Repository.dart';

import '../Domain/Activity.dart';

class ActivitiesService {
  Repository activitiesRepo;

  ActivitiesService(this.activitiesRepo);

  List<Activity> getActivities() {
    return activitiesRepo.getActivities();
  }

  Activity getActivity(int id) {
    return activitiesRepo.getActivity(id);
  }

  bool isOverlapping(TimeOfDay newStartTime, TimeOfDay newEndTime, TimeOfDay existingStartTime, TimeOfDay existingEndTime) {
    final newStartMinutes = newStartTime.hour * 60 + newStartTime.minute;
    final newEndMinutes = newEndTime.hour * 60 + newEndTime.minute;

    final existingStartMinutes = existingStartTime.hour * 60 + existingStartTime.minute;
    final existingEndMinutes = existingEndTime.hour * 60 + existingEndTime.minute;

    return newStartMinutes < existingEndMinutes && newEndMinutes > existingStartMinutes;
  }

  void checkOverlap(Activity newActivity) {
      for (Activity activity in getActivities()) {
        if (newActivity.date == activity.date && newActivity.id != activity.id) {
          if (isOverlapping(newActivity.startTime, newActivity.endTime, activity.startTime, activity.endTime)) {
            throw Exception("Activity overlaps with an existing activity");
          }
        }
      }
  }

  void checkInvalidTime(Activity activity) {
    final startTimeMins = activity.startTime.hour * 60 + activity.startTime.minute;
    final endTimeMins = activity.endTime.hour * 60 + activity.endTime.minute;
    if (startTimeMins > endTimeMins) {
      throw Exception("Activity's start time is after end time");
    }
  }

  void addActivity(Activity activity) {
    if (activity.title == "") {
      throw Exception("Insert a title for this activity");
    }
    checkOverlap(activity);
    checkInvalidTime(activity);
    activitiesRepo.addActivity(activity);
  }

  void updateActivity(Activity activity) {
    if (activity.title == "") {
      throw Exception("Insert a title for this activity");
    }
    checkOverlap(activity);
    checkInvalidTime(activity);
    activitiesRepo.updateActivity(activity);
  }

  void removeActivity(Activity activity) {
    activitiesRepo.removeActivity(activity);
  }
}